### PR TITLE
feat(medusa): Allow tax-inclusive pricing updates on relevant entities

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

The following entity endpoints should allow to create or update the related entity with `includes_tax` field
- Region
- PriceList
- ShippingOption
- currency

**Currency**
The same should be allowed for the currency but it also require to create a new service and new endpoints (get/list/update) to support that feature.

**Feature flag**
!I did fix it in that pr in order to be able to to the tests suite as expected! But it could be move into a separate pr since it is not related to the scope of that pr. Will await till we decide.

Fix the feature flag management. The container contains a flagRouter, but the decorators reload the router each time they are executed. In the end, in the unit tests, if you set a flag, the decorators are not aware of your values since it reload them from the loader
What I did is that I have created a singleton that is used by the loader to set the flag properly and that can be also consume by the decorators without having to reload them each time through the loader which can create some inconsistencies

FIXES CORE-381

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/currency/list-currencies.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * summary: "List Currency"
+ * description: "Retrieves a list of Currency"
+ * x-authenticated: true
+ * parameters:
+ *   - (query) q {string} Query used for searching currencies.
+ *   - (query) code {string} Code of the currency to search for.
+ *   - (query) includes_tax {boolean} Search for tax inclusive currencies.
+ *   - (query) order {string} to retrieve products in.
+ *   - (query) offset {string} How many products to skip in the result.
+ *   - (query) limit {string} Limit the number of products returned.
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/currency/list-currencies.ts:19]: todo: Does not match allowed params. `q`, `code`, and `includes_tax` seems to be missing from `AdminGetCurrenciesParams`

File: packages/medusa/src/api/routes/admin/price-lists/create-price-list.ts [hunk 2]
```diff
@@ -39,6 +42,9 @@ import { EntityManager } from "typeorm"
  *             enum:
  *              - sale
  *              - override
+ *        includes_tax:
+ *             description: "[EXPERIMENTAL] Is the price list tax inclusive or not"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/price-lists/create-price-list.ts:46]: suggestion: Rephrase to improve semantics
```suggestion
 *             description: "[EXPERIMENTAL] Tax included in prices of price list"
```

File: packages/medusa/src/api/routes/admin/price-lists/update-price-list.ts [hunk 3]
```diff
@@ -81,6 +84,9 @@ import { EntityManager } from "typeorm"
  *                 id:
  *                   description: The id of a customer group
  *                   type: string
+ *        includes_tax:
+ *             description: "[EXPERIMENTAL] Is the price list tax inclusive or not"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/price-lists/update-price-list.ts:88]: suggestion: Rephrase to improve semantics
```suggestion
 *             description: "[EXPERIMENTAL] Tax included in prices of price list"
```

File: packages/medusa/src/api/routes/admin/regions/create-region.ts [hunk 4]
```diff
@@ -46,6 +54,9 @@ import { EntityManager } from "typeorm"
  *             type: array
  *             items:
  *               type: string
+ *         includes_tax:
+ *             description: "[EXPERIMENTAL] Is the region tax inclusive or not"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/regions/create-region.ts:58]: suggestion: Rephrase to improve semantics
```suggestion
 *             description: "[EXPERIMENTAL] Tax included in prices in region"
```

File: packages/medusa/src/api/routes/admin/regions/update-region.ts [hunk 5]
```diff
@@ -44,6 +46,9 @@ import { EntityManager } from "typeorm"
  *           tax_rate:
  *             description: "The tax rate to use on Orders in the Region."
  *             type: number
+ *          includes_tax:
+ *             description: "[EXPERIMENTAL] Is the region tax inclusive or not"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/regions/update-region.ts:50]: suggestion: Rephrase to improve semantics
```suggestion
 *             description: "[EXPERIMENTAL] Tax included in prices in region"
```

File: packages/medusa/src/api/routes/admin/shipping-options/create-shipping-option.ts [hunk 6]
```diff
@@ -69,6 +72,9 @@ import { EntityManager } from "typeorm"
  *           metadata:
  *             description: An optional set of key-value pairs with additional information.
  *             type: object
+ *          includes_tax:
+ *             description: "[EXPERIMENTAL] Does the shipping option includes tax"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/shipping-options/create-shipping-option.ts:76]: Same as other comments. Will not add for rest

**Review by olivermrbl — COMMENTED** [review]
Great work! Have left a couple of comments here and there 💪

File: integration-tests/api/__tests__/admin/currency.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    })
+
+    it("should retrieve the currencies", async () => {
+      const api = useApi()
+      const response = await api.get(
+        `/admin/currencies?order=code`,
+        adminReqConfig
+      )
+
+      expect(response.data).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply integration-tests/api/__tests__/admin/currency.js:58]: yes the snapshot is uniq and is composed of the `describes` and `its` description

File: packages/medusa/src/services/currency.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      }
+
+      const currencyRepo = transactionManager.getCustomRepository(
+        this.currencyRepository_
+      )
+      await currencyRepo.save(currency)
+
+      await this.eventBusService_.emit(CurrencyService.EVENTS.UPDATE, {
+        code,
+      })
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa/src/services/currency.ts:126]: Yeah could make sense to update prices through a subscriber 👍

File: packages/medusa/src/services/currency.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+type InjectedDependencies = {
+  manager: EntityManager
+  currencyRepository: typeof CurrencyRepository
+  eventBusService: EventBusService
+  featureFlagRouter: FlagRouter
+}
+
+export default class CurrencyService extends TransactionBaseService<CurrencyService> {
+  static readonly EVENTS = {
+    UPDATE: "currency.update",
```
Comment (srindom) [inline packages/medusa/src/services/currency.ts:22]: todo: 

```suggestion
    UPDATED: "currency.updated",
```

File: packages/medusa/src/services/currency.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+type InjectedDependencies = {
+  manager: EntityManager
+  currencyRepository: typeof CurrencyRepository
+  eventBusService: EventBusService
+  featureFlagRouter: FlagRouter
+}
+
+export default class CurrencyService extends TransactionBaseService<CurrencyService> {
+  static readonly EVENTS = {
```
Comment (srindom) [inline packages/medusa/src/services/currency.ts:21]: nit: maybe just `Events` as in other places?

File: packages/medusa/src/services/currency.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+   * @return The currency
+   */
+  async retrieve(code: string): Promise<Currency | never> {
+    const currencyRepo = this.manager_.getCustomRepository(
+      this.currencyRepository_
+    )
+
+    const currency = await currencyRepo.findOne({
+      where: { code },
+    })
```
Comment (srindom) [inline packages/medusa/src/services/currency.ts:57]: thought: to make the retrieve API easier to work with would it makes sense to normalize the code?

```
const code = rawCode.toLowerCase()
```

Comment (adrien2p) [reply packages/medusa/src/services/currency.ts:57]: a lot yeah

Comment (olivermrbl) [reply packages/medusa/src/services/currency.ts:58]: Also, would it make sense to rename the method to `retrieveByCode` to differentiate it from "normal" `retrieve` methods that always expect an `id`?

Comment (adrien2p) [reply packages/medusa/src/services/currency.ts:58]: good idea 👍

**Comment by olivermrbl** [comment]
@adrien2p Unit tests need to be updated to reflect the normalization of currency code

**Pull request merged**